hammer/hammer.go

```diff
@@ -185,6 +193,48 @@ func (h *Hammer) Run(ctx context.Context) {
 			}
 		}
 	}()
+
+	go func() {
+		tick := time.NewTicker(100 * time.Millisecond)
+		size := h.tracker.LatestConsistent.Size
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-tick.C:
+			}
+			newSize := h.tracker.LatestConsistent.Size
+			if newSize <= size {
+				continue
+			}
+			now := time.Now()
+			totalLatency := time.Duration(0)
+			queueLatency := time.Duration(0)
+			numLeaves := 0
+			var sample *leafTime
+			for {
+				if sample == nil {
+					l, ok := <-h.leafSampleChan
+					if !ok {
+						break
+					}
+					sample = &l
+				}
+				if sample.idx >= newSize || sample.assignedAt.After(now) {
+					break
+				}
+				queueLatency += sample.assignedAt.Sub(sample.queuedAt)
+				totalLatency += now.Sub(sample.queuedAt)
+
+				numLeaves++
+				sample = nil
+			}
+			if numLeaves > 0 {
+				h.integrationTime.Add(float64(totalLatency/time.Millisecond) / float64(numLeaves))
+				h.queueTime.Add(float64(queueLatency/time.Millisecond) / float64(numLeaves))
```

Comment:
Is there a reason to do it this way instead of posting the row values? This is a bit blurry because it calculates a mean here, and then the moving average will calculate averages from them (if I understand the library right).

Reply:
Mainly just trying to keep the size of the retained data down, and have some sort of loose idea of how the window relates to time even as throughput varies.

I believe that a mean of means is the same as mean of all the things _if_ the subsets are of equal size. Admittedly this is _probably_ not always the case here (certainly if throughput is varying and < 100qps then it's pretty unlikely to be), but tbh with the sampling and "observed" nature of this it's probably all only really good enough for a rough guide to what's going on, but that should be enough to tell you if your log is melting and falling further and further behind, for example.

I'm sure there are better/more precise ways of doing it :)